python/lsst/ap/verify/workspace.py

```diff
@@ -0,0 +1,82 @@
+#
+# LSST Data Management System
+# Copyright 2017 LSST Corporation.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+from __future__ import absolute_import, division, print_function
+
+import os
+import stat
+
+
+class Workspace(object):
+    """A directory used by ``ap_verify`` to handle data.
+
+    Any object of this class represents a plan for organizing a working
+    directory. At present, constructing a Workspace does not initialize its
+    repositories; for compatibility reasons, this is best deferred to
+    individual tasks. This class may later provide Butler objects if this
+    becomes a useful feature.
+
+    Parameters
+    ----------
+    location : `str`
+       The location on disk where the workspace will be set up. Will be
+       created if it does not already exist.
+
+    Raises
+    ------
+    `EnvironmentError`
+        ``location`` is not readable or not writeable
+    """
+
+    def __init__(self, location):
+        # Can't use exceptions to reliably distinguish existing directory from other cases in Python 2
+        if os.path.isdir(location):
+            if not os.access(location, os.R_OK | os.W_OK):
+                raise IOError('Workspace % cannot be read.' % location)
+        else:
+            os.makedirs(location, stat.S_IRWXU | stat.S_IRGRP | stat.S_IROTH)
+
+        self._location = location
+
+    @property
+    def dataRepo(self):
+        """The URI to a Butler repo for science data (`str`, read-only).
+        """
+        return os.path.join(self._location, 'ingested')
```

Comment:
(Here and below)

Being picky, if you really mean URI, is it a good idea to use `os.sep` to join? I imagine it's totally fine on any LSST-supported platfom, but maybe it should really be `urlparse.urljoin`?

Reply:
A lot of the existing code assumes these are filesystem paths (see [DM-11482](https://jira.lsstcorp.org/browse/DM-11482)). The problem with just using `urlparse` is that what the Butler calls a URI can really be *either* a URI (by the conventional definition) or a filesystem path, so I'd need to be able to handle both.

TL;DR: I suspect that a dumb replacement would break existing code, so I'm deferring the fix to DM-11482.